torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+                    fn_counts.append((ir, file_function))
+                    continue
+
+                if begin_collecting and re.match(r"-+", l):
+                    continue
+
+                begin_collecting = False
+
+            return fn_counts
+        finally:
+            shutil.rmtree(working_dir)
+
+    @staticmethod
+    def _construct_script(stmt: str, setup: str, number: int, num_threads: int, error_log: str):
+        return textwrap.dedent(r"""
+            import os
+            import subprocess
+            import sys
+
+            import torch
+            torch.set_num_threads({num_threads})
+
+            PID = os.getpid()
+
+            def log_failure(msg):
+                with open("{error_log}", "wt") as f:
+                    f.write(msg)
+                sys.exit(1)
+
+            def check_result(completed_process):
+                if completed_process.returncode:
+                    log_failure(f"Command failed: {{' '.join(completed_process.args)}}")
+                return completed_process
+
+            # =============================================================================
+            # == Check that subprocess matches parent =====================================
+            # =============================================================================
+            if sys.executable != "{parent_interpreter}":
+                log_failure(
+                    "Interpreter mismatch:\n"
+                    f"  {{sys.executable}}\n    vs.\n  {parent_interpreter}"
+                )
+
+            if torch.__file__ != "{torch_file}":
+                log_failure(
+                    "PyTorch does not match expected file:\n"
+                    f"  {{torch.__file__}}\n    vs.\n  {torch_file}"
+                )
+
+            # =============================================================================
+            # == User specified setup =====================================================
+            # =============================================================================
+            {setup}
+
+            for _ in range({warmup_number}):
+                {stmt}
+
+            # =============================================================================
+            # == Callgrind management =====================================================
+            # =============================================================================
+            callgrind_stat = check_result(subprocess.run(
+                ["callgrind_control", "--stat"],
+                capture_output=True,
+            ))
+
+            stat_lines = callgrind_stat.stdout.decode("utf-8").splitlines()
+            if f"PID {{PID}}: python {{__file__}}" not in stat_lines:
+                log_failure("Process does not appear to be running callgrind.")
+
+            # Disable and zero counters to ensure a clean run.
+            for cmd in ["--instr=off", "--zero", "--instr=on"]:
```

Comment:
So is the noise in measurements coming mostly from the fact that we subprocess here and thus measure some instructions involved in waiting for the process's output?

I wonder whether we could use the macro instruction itself: CALLGRIND_START_INSTRUMENTATION . I think we could even bake it in PyTorch build itself by building a tiny side library (or putting it into libtorch.so). Basically that macro just desugars into a magic asm instructions that callgrind recognizes: https://github.com/rantoniello/valgrind/blob/21be3371beb74cff971e0bbf48fbecc7bf970e49/include/valgrind.h#L269 - there doesn't seem to be any static/dynamic dependency on valgrind. So we could just build with it unconditionally or if needed - copy the macro implementation separately.

Reply:
I did the dumb thing and just copied `callgrind.h` and `valgrind.h` to bind the macros, and the variation almost completely disappeared. The remaining stragglers for a trivial workload are some Python and pybind calls / destructors which are generally single digit number of instructions and `work/Objects/dictobject.c:lookdict_unicode_nodummy` which is generally hundreds to low thousands.

On a test of `torch.ones((1,))`, jitter (which I define as fewer operations than the "no-op" reference run that Timer does) was 1146 instructions compared to 1.1 million "real" instructions for `number=100`, and 1492 vs. 11 million for `number=1000`. (`number` is iteration count) Adding `gc.collect()` statements seems to clear up that noise, but replaced it with some jitter from `gc` functions.

It's not perfect, but it's something like 100x less noise than the subprocess control approach, and I'm skeptical that it's possible to squeeze Python any further. I'll push something tomorrow and then solicit advice on how to get the macros into the build in a more principled way.

Reply:
I did a bit more tinkering with `gc` as well as a short sleep (10 ms) to let everything flush through the interpreter and that trimmed the jitter down even more.